akka-http-caching/src/main/scala/akka/http/caching/LruCache.scala

```diff
@@ -0,0 +1,92 @@
+package akka.http.caching
+
+import java.util.concurrent.{ CompletableFuture, Executor, TimeUnit }
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+import scala.concurrent.Future
+import com.github.benmanes.caffeine.cache.{ AsyncCacheLoader, AsyncLoadingCache, Caffeine }
+import akka.http.caching.LruCache.toJavaMappingFunction
+
+import scala.compat.java8.FutureConverters._
+import scala.compat.java8.FunctionConverters._
+
+object LruCache {
+
+  /**
+   * Creates a new [[akka.http.caching.ExpiringLruCache]] or
+   * [[akka.http.caching.SimpleLruCache]] instance depending on whether
+   * a non-zero and finite timeToLive and/or timeToIdle is set or not.
+   */
+  def apply[V](
+    defaultLoader:   Any ⇒ Future[V],
+    maxCapacity:     Int             = 500,
+    initialCapacity: Int             = 16,
+    timeToLive:      Duration        = Duration.Inf,
+    timeToIdle:      Duration        = Duration.Inf): Cache[V] = {
+
+    val loader = new AsyncCacheLoader[Any, V] {
+      def asyncLoad(k: Any, e: Executor) = defaultLoader(k).toJava.toCompletableFuture
+    }
+
+    if (timeToLive.isFinite() || timeToIdle.isFinite())
+      new ExpiringLruCache[V](maxCapacity, initialCapacity, timeToLive, timeToIdle, loader)
+    else
+      new SimpleLruCache[V](maxCapacity, initialCapacity, loader)
+  }
+
+  def toJavaMappingFunction[V](genValue: () ⇒ Future[V]) =
+    asJavaBiFunction[Any, Executor, CompletableFuture[V]]((k, e) ⇒ genValue().toJava.toCompletableFuture)
+}
+
+trait LruCache[V] extends Cache[V] {
+
+  private[caching] val store: AsyncLoadingCache[Any, V]
+
+  def get(key: Any): Option[Future[V]] = Option(store.getIfPresent(key)).map(_.toScala)
+
+  def apply(key: Any, genValue: () ⇒ Future[V]): Future[V] = store.get(key, toJavaMappingFunction(genValue)).toScala
+
+  def remove(key: Any): Option[Future[V]] = Option(store.synchronous().asMap().remove(key)).map(Future.successful)
```

Comment:
The spray implementation does not use `remove`, and I can't think why it would be useful to return the value. I had tried to keep the definitions consistent, but it's tempting to change this to:
`def remove(key: Any): Unit`